File: basic-app/src/main/java/com/github/vincentfree/clients/OrderClient.java
package com.github.vincentfree.clients;

import com.github.vincentfree.model.Order;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.TimerTask;
import java.util.UUID;
import java.util.concurrent.ThreadLocalRandom;

public class OrderClient {
    private final HttpClient client;
    private String[] words;

    public OrderClient(HttpClient client) {
        this.client = client;
        String wordAPIUrl = "https://random-word-api.herokuapp.com/word?number=150";
        var request = HttpRequest.newBuilder(URI.create(wordAPIUrl)).GET()
                .timeout(Duration.ofSeconds(5))
                .version(HttpClient.Version.HTTP_1_1)
                .build();
        try {
            var response = client.send(request, HttpResponse.BodyHandlers.ofString());
            words = response.body()
                    .replace("[", "")
                    .replace("]", "")
                    .split(",");
        } catch (IOException | InterruptedException err) {
            System.err.printf("failed to get array of words..%n err: %s", err.getMessage());
            words = new String[]{
                    "uppercutting",
                    "appreciation",
                    "cyclothymias",
                    "mismanagements",
                    "parthenogenetic",
                    "pottle",
                    "maras"
            };
        }
    }

    private HttpRequest generateOrderRequest() {
        var word = words[ThreadLocalRandom.current().nextInt(words.length)];
        var order = new Order(word, word, UUID.randomUUID().toString());
        var payload = wrapPayload(order.toJsonString());
        return HttpRequest.newBuilder()
                .POST(HttpRequest.BodyPublishers.ofString(payload))
                .uri(URI.create("http://localhost:3500/v1.0/publish/kafka/order.events"))
                .version(HttpClient.Version.HTTP_1_1)
                .build();
    }

    private HttpResponse<String> sendEvent(HttpRequest request) throws IOException, InterruptedException {
        return client.send(request, HttpResponse.BodyHandlers.ofString());
    }

    public TimerTask orderTimerTask() {
        return new TimerTask() {
            public void run() {
                try {
                    System.out.println("Sending event to kafka..");
                    var result = sendEvent(generateOrderRequest());
                    System.out.printf("The status is: %d%n", result.statusCode());
                } catch (IOException | InterruptedException err) {
                    System.err.println("Failed to send event to kafka!");
                    throw new RuntimeException(err);
                }
            }
        };
    }

    protected String wrapPayload(String payload) {
        return "{ \"data\": {" + payload + "} }";
    }
}

File: basic-app-go/order/orderClient.go
package order

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"time"

	"github.com/dapr/go-sdk/client"
	"github.com/google/uuid"
)

var (
	words = [7]string{
		"uppercutting",
		"appreciation",
		"cyclothymias",
		"mismanagements",
		"parthenogenetic",
		"pottle",
		"maras",
	}
	msgChannel = make(chan string)
	logger     = log.Default()
)

type DaprClient struct {
	Dapr client.Client
}

type Order struct {
	Id      string `json:"id"`
	Name    string `json:"name"`
	OrderId string `json:"orderId"`
}

func (o *Order) toJson() ([]byte, error) {
	jsonBody, err := json.Marshal(o)
	if err != nil {
		return nil, err
	}
	return jsonBody, nil

}

func (o *Order) ToJsonString() string {
	order, err := o.toJson()
	if err != nil {
		return "{\"message\": \"Failed to decode json\"}"
	}
	return string(order)
}

func (c *DaprClient) request() {
	ctx := context.Background()
	word := words[rand.Intn(len(words))]
	randomUUID, _ := uuid.NewRandom()
	o := Order{Id: word, Name: word, OrderId: randomUUID.String()}
	b, _ := o.toJson()
	err := c.Dapr.PublishEvent(ctx, "pubsub", "order.events", b)
	if err != nil {
		fmt.Printf("Failed to send order event: %s to kafka\n", string(b))
	}
	msgChannel <- fmt.Sprintf("the order message: %s, has been send", o.ToJsonString())
}

func (c *DaprClient) OrderTimerTask(d time.Duration) {
	tick := time.NewTicker(d)
	for {
		select {
		case <-tick.C:
			go c.request()
		case resp := <-msgChannel:
			logger.Printf("The published message: %s\n", resp)
		}
	}
}
